## Sample Data Files <br>

### Title: A three-dimensional map of the human genome at kilobase resolution reveals prinicples of chromatin looping <br>
File: K562_NHEK_GM12878-loops.txt <br>
Summary: We use in situ Hi-C to probe the three-dimensional architecture of genomes, constructing haploid and diploid maps of nine cell types. The densest, in human lymphoblastoid cells, contains 4.9 billion contacts, achieving 1-kilobase resolution. We find that genomes are partitioned into local domains, which are associated with distinct patterns of histone marks and segregate into six subcompartments. We identify ~10,000 loops. These loops frequently link promoters and enhancers, correlate with gene activation, and show conservation across cell types and species. Loop anchors typically occur at domain boundaries and bind CTCF. CTCF sites at loop anchors occur predominantly (>90%) in a convergent orientation, with the asymmetric motifs 'facing' one another. The inactive X-chromosome splits into two massive domains and contains large loops anchored at CTCF-binding repeats. <br>
Organism: Human <br>

Link to GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63525 <br>

Publications: <br>

Rao, Suhas S.P., et al. A 3D Map of the Human Genome at Kilobase Resolution Reveals Principles of Chromatin Looping. Cell 2014;159(7):1665-1680.<br>

Sanborn, A.L., et al. Chromatin extrusion explains key features of loop and domain formation in wild-type and engineered genomes. Proceedings of the National Academy of Sciences 2015;112(47):E6456-E6465.<br>

### Title: Capture Hi-C reveals novel candidate genes and complex long-range interactions with related autoimmune risk loci
File: GSM1704495_GMPro_Cap_rep1_filt5.bedpe, GSM1704494_JKProCap_Rep1_Filt5.bedpe <br>
Summary: Genome-wide association studies have been tremendously successful in identifying genetic variants associated with complex diseases. The majority of association signals are intergenic and evidence is accumulating that a high proportion of signals lie in enhancer regions.We use Capture Hi-C to investigate, for the first time, the interactions between associated variants for four autoimmune diseases and their functional targets in B- and T-cell lines. Here we report numerous looping interactions and provide evidence that only a minority of interactions are common to both B- and T-cell lines, suggesting interactions may be highly cell-type specific; some disease-associated SNPs do not interact with the nearest gene but with more compelling candidate genes (for example, FOXO1, AZI2) often situated several megabases away; and finally, regions associated with different autoimmune diseases interact with each other and the same promoter suggesting common autoimmune gene targets (for example, PTPRC, DEXI and ZFP36L1). <br>

Link to GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE69600 <br>

Publications: <br>

Martin, P., et al. Identifying Causal Genes at the Multiple Sclerosis Associated Region 6q23 Using Capture Hi-C. PLOS ONE 2016;11(11):e0166923.<br>

Martin, P., et al. Capture Hi-C reveals novel candidate genes and complex long-range interactions with related autoimmune risk loci. Nature Communications 2015;6:10069.<br>

###  Title: ImmGen ATAC-seq data
Files: Tcon_Immgen_3.bg Treg_Immgen_3.bg <br>
       Tcon_Immgen_3_chr1.rar <br>

Summary: Immunological Genome Project chromatin accessibility maps for 86 different immunocytes (ATAC-seq). Immune cell populations were isolated in high-purity by flow cytometry.
Overall Design: Immune cell populations encompassing all major lineages were isolated by flow cytometry to high-purity according to ImmGen SOP (https://www.immgen.org/). 10,000 cells (except some rare cell populations) were used to construct each ATAC-seq library, using the Fast-ATAC protocol. Libraries were dual-indexed and pooled for high-throughput sequencing.
Highly purified cell samples were isolated in parallel from the same populations for gene expression profiling by low-input RNAseq (ImmGen protocol, per umbrella GSE126267). These corresponding RNAseq datasets are available at GSE109125.

Link to GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE100738

Publications: 
Yoshida, H., Lareau, C. A., Ramirez, R. N., Rose, S. A., Maier, B., Wroblewska, A., . . . Benoist, C. (2019). The cis-Regulatory Atlas of the Mouse Immune System. Cell, 176(4), 897-912.e820. doi:https://doi.org/10.1016/j.cell.2018.12.036

These files are too large to include here so we are including a link to the google cloud instance:<br>
[`Tcon_Immgen_3.bg`](https://storage.googleapis.com/gencode_ch_data/Sample_Data/datasets/Tcon_Immgen_3.bg)<br> 
[`Treg_Immgen_3.bg`](https://storage.googleapis.com/gencode_ch_data/Sample_Data/datasets/Treg_Immgen_3.bg)<br>

###  Title: FoxP3_5047 data
Files: FoxP3_5047_Peaks.txt
FoxP3_5047_Peaks - no headers.txt

Abstract: FoxP3 conditions the transcriptional signature and functional facets of regulatory T cells (Treg cells). Its mechanism of action, whether as an activator or a repressor, has remained unclear. Here, chromatin analysis showed that FoxP3 bound active enhancer elements, not repressed chromatin, around loci over- or under-expressed in Treg cells. We evaluated the impact of a panel of FoxP3 mutants on its transcriptional activity and interactions with DNA, transcriptional cofactors and chromatin. Computational integration, confirmed by biochemical interaction and size analyses, showed that FoxP3 existed in distinct multimolecular complexes. It was active and primarily an activator when complexed with the transcriptional factors RELA, IKZF2 and KAT5. In contrast, FoxP3 was inactive when complexed with the histone methyltransferase EZH2 and transcription factors YY1 and IKZF3. The latter complex partitioned to a peripheral region of the nucleus, as shown by super-resolution microscopy. Thus, FoxP3 acts in multimodal fashion to directly activate or repress transcription, in a context- and partner-dependent manner, to govern Treg cell phenotypes.Link to GEO: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5679728/
Publications: Kwon, H.-K., Chen, H.-M., Mathis, D., & Benoist, C. (2017). Different molecular complexes that mediate transcriptional induction and repression by FoxP3. Nature Immunology, 18(11), 1238-1248. doi:10.1038/ni.3835


